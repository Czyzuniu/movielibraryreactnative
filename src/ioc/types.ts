const InjectableTypes = {
  GetPopularMoviesUseCase: Symbol.for('GetPopularMoviesUseCase'),
  FindMovieByTextUseCase: Symbol.for('FindMovieByTextUseCase'),
  GetMovieByIdUseCase: Symbol.for('GetMovieByIdUseCase'),
  MovieRepository: Symbol.for('MovieRepository'),
  MovieRemoteDataSource: Symbol.for('MovieRemoteDataSource'),
  AxiosClient: Symbol.for('AxiosClient'),
  AsyncStorageRepository: Symbol.for('AsyncStorageRepository'),
};

export {InjectableTypes};
