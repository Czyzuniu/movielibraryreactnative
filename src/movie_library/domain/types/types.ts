export enum ProductCategoryType {
  Electronics = 'electronics',
  Jewelery = 'jewelery',
  ClothesWomen = "women's clothing",
  ClothesMen = "men's clothing",
}

export interface IUseCase<P, R> {
  execute: (params: P) => R;
}
