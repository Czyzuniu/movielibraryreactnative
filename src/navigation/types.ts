import {StackNavigationProp} from "@react-navigation/stack";
import {BottomTabNavigationProp} from "@react-navigation/bottom-tabs";

type PopularMoviesStackParamList = {
  PopularMovies: undefined,
  Common: StackNavigationProp<CommonStackParamList>
}

type CreatedListsStackStackParamList = {
  MyLists: undefined,
  CreateList: undefined,
  Common: StackNavigationProp<CommonStackParamList>
}

type SettingsStackTypeParamList = {
  Settings: undefined
}

type LoggedInBottomNavigationTabParamList = {
  PopularMoviesStack: StackNavigationProp<PopularMoviesStackParamList>;
  CreatedListsStack: StackNavigationProp<CreatedListsStackStackParamList>;
  SettingsStack: StackNavigationProp<SettingsStackTypeParamList>;
};

type LoginStackParamList = {
  Login: undefined
}

type RootStackParamList = {
  LoggedOut: StackNavigationProp<LoginStackParamList>
  LoggedIn: BottomTabNavigationProp<LoggedInBottomNavigationTabParamList>;
};

type CommonStackParamList = {
  ViewMovie: {movieId: string; title: string};
  ViewMovieHomePageWebView: {movieUrl: string};
};

export {
  LoggedInBottomNavigationTabParamList,
  RootStackParamList,
  CommonStackParamList,
  PopularMoviesStackParamList,
  SettingsStackTypeParamList,
  LoginStackParamList,
  CreatedListsStackStackParamList
};
