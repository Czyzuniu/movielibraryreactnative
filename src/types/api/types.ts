type AuthTokenResponse = {
  success: boolean,
  expires_at: string,
  request_token: string
}

type AuthNewSessionResponse = {
  success: boolean,
  session_id: string
}

type AddToFavouritesRequestBody = {
  accountId: string,
  movieId: number,
  type: 'movie' | 'tv'
}

type AccountDetailsResponse = {
  "id": number,
  "name": string,
  "include_adult": boolean,
  "username": string
}

type LoginRequest = {
  username: string,
  password: string
  request_token: string
}

export { AuthTokenResponse, LoginRequest, AuthNewSessionResponse, AccountDetailsResponse, AddToFavouritesRequestBody}
