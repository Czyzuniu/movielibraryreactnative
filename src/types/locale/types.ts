export type LocaleSettings = {
  language: string
}
