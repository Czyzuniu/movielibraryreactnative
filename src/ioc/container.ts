import 'reflect-metadata';
import {Container, ContainerModule} from 'inversify';
import AxiosClient from '../base/infrastructure/AxiosClient';
import {InjectableTypes} from './types';
import {
  MoviesRemoteApiDataSource,
  MoviesRemoteApiDataSourceImpl,
} from '../movie_library/infrastructure/datasource/MoviesRemoteApiDataSource';
import {MovieRepository} from '../movie_library/domain/repository/MovieRepository';
import MovieRepositoryImpl from '../movie_library/infrastructure/repository/MovieRepositoryImpl';
import GetPopularMoviesUseCase from '../movie_library/domain/usecase/GetPopularMoviesUseCase';
import GetMovieByIdUseCase from '../movie_library/domain/usecase/GetMovieByIdUseCase';
import {AsyncStorageRepository} from '../async_storage/domain/repository/AsyncStorageRepository';
import AsyncStorageRepositoryImpl from '../async_storage/infrastructure/AsyncStorageRepositoryImpl';

const myContainer = new Container();

const useCases = new ContainerModule(bind => {
  bind<GetPopularMoviesUseCase>(InjectableTypes.GetPopularMoviesUseCase).to(
    GetPopularMoviesUseCase,
  );
  bind<GetMovieByIdUseCase>(InjectableTypes.GetMovieByIdUseCase).to(
    GetMovieByIdUseCase,
  );
});

const repositories = new ContainerModule(bind => {
  bind<MovieRepository>(InjectableTypes.MovieRepository)
    .to(MovieRepositoryImpl)
    .inSingletonScope();

  bind<AsyncStorageRepository>(InjectableTypes.AsyncStorageRepository)
    .to(AsyncStorageRepositoryImpl)
    .inSingletonScope();
});

const datasources = new ContainerModule(bind => {
  bind<MoviesRemoteApiDataSource>(InjectableTypes.MovieRemoteDataSource)
    .to(MoviesRemoteApiDataSourceImpl)
    .inSingletonScope();
});

const clients = new ContainerModule(bind => {
  bind<AxiosClient>(InjectableTypes.AxiosClient)
    .to(AxiosClient)
    .inSingletonScope();
});

myContainer.load(useCases, repositories, datasources, clients);

export {myContainer};
