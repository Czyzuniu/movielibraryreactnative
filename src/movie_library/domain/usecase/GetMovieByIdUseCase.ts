import {MovieRepository} from '../repository/MovieRepository';
import {InjectableTypes} from '../../../ioc/types';
import {inject, injectable} from 'inversify';
import DescriptiveMovie from '../entity/DescriptiveMovie';

@injectable()
export default class GetMovieByIdUseCase {
  private repository: MovieRepository;

  constructor(
    @inject(InjectableTypes.MovieRepository) repository: MovieRepository,
  ) {
    this.repository = repository;
  }

  async execute(movieId: string): Promise<DescriptiveMovie> {
    return this.repository.getMovieById(movieId);
  }
}
