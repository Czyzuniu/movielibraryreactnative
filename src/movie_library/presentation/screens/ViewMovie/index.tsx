import React, {useEffect} from 'react';
import {ScrollView} from 'native-base';
import {StackScreenProps} from '@react-navigation/stack';
import {HomeStackParamList} from '../../../../navigation/types';
import {myContainer} from '../../../../ioc/container';
import {InjectableTypes} from '../../../../ioc/types';
import {useQuery} from 'react-query';
import {MovieLibraryQueryKey} from '../../../consts/consts';
import GetMovieByIdUseCase from '../../../domain/usecase/GetMovieByIdUseCase';
import MovieDetails from '../../components/MovieDetails';
import {Alert, RefreshControl} from 'react-native';

type Props = StackScreenProps<HomeStackParamList, 'ViewMovie'>;

export default function ViewMovie({navigation, route}: Props) {
  const getMovieByIdUseCase = myContainer.get<GetMovieByIdUseCase>(
    InjectableTypes.GetMovieByIdUseCase,
  );

  const {movieId} = route.params;
  const {data, isFetching, refetch} = useQuery(
    `${MovieLibraryQueryKey.GetMovieById}_${movieId}`,
    () => getMovieByIdUseCase.execute(movieId),
    {
      onError: (e: any) => Alert.alert(e.message),
      retry: 0,
    },
  );

  useEffect(() => {
    navigation.setOptions({
      headerTitle: data?.title,
    });
  }, [navigation, data]);

  return (
    <ScrollView
      testID={'VIEW_MOVIE_SCROLL'}
      refreshControl={
        <RefreshControl refreshing={isFetching} onRefresh={refetch} />
      }>
      {data && <MovieDetails movie={data} />}
    </ScrollView>
  );
}
