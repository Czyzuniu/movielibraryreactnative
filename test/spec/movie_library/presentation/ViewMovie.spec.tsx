import {render, waitFor} from '@testing-library/react-native';
import React from 'react';
import ContextWrapper from '../../../utils/ContextWrapper';
import {mockNavigation} from '../../../utils/mockNavigation';
import {mockDeep} from 'jest-mock-extended';
import {when} from 'jest-when';
import {myContainer} from '../../../../src/ioc/container';
import {InjectableTypes} from '../../../../src/ioc/types';
import GetMovieByIdUseCase from '../../../../src/movie_library/domain/usecase/GetMovieByIdUseCase';
import ViewMovie from '../../../../src/movie_library/presentation/screens/ViewMovie';
import TestObjects from '../../../utils/TestObjects';
import {Alert} from 'react-native';
import {act} from 'react-test-renderer';

describe('View Movie Screen Tests', () => {
  let getMovieByIdUseCase: GetMovieByIdUseCase;

  beforeEach(() => {
    getMovieByIdUseCase = mockDeep<GetMovieByIdUseCase>();
    jest.spyOn(myContainer, 'get').mockImplementation(() => jest.fn());
    when(myContainer.get)
      .calledWith(InjectableTypes.GetMovieByIdUseCase)
      .mockReturnValue(getMovieByIdUseCase);
  });

  it('should render and match snapshot', () => {
    const {toJSON} = render(
      <ContextWrapper>
        <ViewMovie {...mockNavigation({movieId: '12345'})} />
      </ContextWrapper>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should call the usecase to fetch the movie initially and render movie details', async () => {
    const movieId = '12345';
    when(getMovieByIdUseCase.execute).mockResolvedValue(
      TestObjects.DescriptiveMovie(),
    );

    const {getByTestId} = render(
      <ContextWrapper>
        <ViewMovie {...mockNavigation({movieId})} />
      </ContextWrapper>,
    );

    await waitFor(async () => {
      expect(getMovieByIdUseCase.execute).toBeCalledWith(movieId);
      expect(
        getByTestId(`MOVIE_DETAILS_${TestObjects.DescriptiveMovie().id}`),
      ).toBeTruthy();
    });
  });

  it('should refetch when pulled to refresh on main scrollview', async () => {
    const movieId = '12345';
    when(getMovieByIdUseCase.execute).mockResolvedValue(
      TestObjects.DescriptiveMovie(),
    );

    const {getByTestId} = render(
      <ContextWrapper>
        <ViewMovie {...mockNavigation({movieId})} />
      </ContextWrapper>,
    );

    const scrollView = getByTestId('VIEW_MOVIE_SCROLL');

    await act(async () => {
      scrollView.props.refreshControl.props.onRefresh();
    });

    expect(getMovieByIdUseCase.execute).toBeCalledWith(movieId);
  });

  it('should display an alert if fails to fetch the movie', async () => {
    jest.spyOn(Alert, 'alert');
    const movieId = '12345';
    const exception = new Error('error');
    const navigation = mockNavigation({movieId});
    when(getMovieByIdUseCase.execute).mockRejectedValue(exception);

    render(
      <ContextWrapper>
        <ViewMovie {...navigation} />
      </ContextWrapper>,
    );

    await waitFor(() => expect(Alert.alert).toBeCalledWith(exception.message));
  });
});
