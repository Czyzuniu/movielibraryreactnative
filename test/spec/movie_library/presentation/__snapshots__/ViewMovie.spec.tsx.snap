// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Movie Screen Tests should render and match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    dataSet={Object {}}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={true}
      />
    }
    style={
      Object {
        "backgroundColor": "white.300",
      }
    }
    testID="VIEW_MOVIE_SCROLL"
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</RNCSafeAreaProvider>
`;
